### PR TITLE
Bugfix/eventbridge/transformer issue with nested key

### PR DESCRIPTION
**Pull request by maxhoheiser**
## Motivation
InputTransformer map values from the event to keys defined in InputPathMap and these can then be used in InputTemplate to replace placeholders.
These placeholders can also be in a continuous string "string/<placeholder>" this was failing in localstack as raised here https://github.com/localstack/localstack/issues/11126

## Changes
- add support for string nested placeholders in json input templates
- add support for string nested placeholders in string input templates

<!--
## Testing
`tests/aws/services/events/test_events_inputs.py::TestInputTransformer::test_input_transformer_nested_keys_replacement`

### CODE REVIEW

**Review by bentsku — CHANGES_REQUESTED** [review]
Nice fix! 

I believe we might still have a few corner cases to cover, we could discover them with some parametrized unit tests for the `is_nested_in_string` function, and maybe even the `replace_template_placeholders` one. 

I think we're not properly replacing `dict` values inside JSON templates string fields, and I'm not 100% certain of the `is_nested_in_string` check if the matched value is between quotes with nothing else, and it'd be great to validate those.

This documentation part is pretty useful to try to understand edge cases: https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-transform-target-input.html#eb-transform-input-issues

I think we should have all the cases in this table: https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-transform-target-input.html#eb-transform-target-input-template as part of unit tests, as they are pretty well defined. Some of those could maybe be tackled in a follow up, but it'd be good to know how far we support. 

Thanks!

File: localstack-core/localstack/services/events/utils.py [hunk 1]
```diff
@@ -273,3 +273,7 @@ def is_nested_in_string(template, match) -> bool:
         return False
 
     return left_quote != -1 and template[left_quote + 1 : right_quote].strip() != match.group(0)
+
+
+def dict_to_simple_string(d):
```
Comment (bentsku) [inline localstack-core/localstack/services/events/utils.py:278]: Ah, I missed this one too: I think this does not cover all use case if there are nested values here. 
It seems what AWS does is stripped all quotes from the JSON string, it might be easier that way.

Because what if you have a nested dict values here? I don't think that works, same if `v` is a list with strings inside.

File: localstack-core/localstack/services/events/target.py [hunk 2]
```diff
@@ -75,23 +77,28 @@ def get_template_replacements(
...
-            return to_json_str(value)
+        if is_json_template:
+            if is_nested_in_string(template, match):
+                return value  # Return raw string for nested placeholders
+            else:
+                return to_json_str(value)
         if isinstance(value, datetime.datetime):
             return event_time_to_time_string(value)
+        if isinstance(value, dict):
+            return dict_to_simple_string(value)
```
[the remark this answers is not in the document]
Comment (maxhoheiser) [reply localstack-core/localstack/services/events/target.py:96]: If I understand you correct this scenario should be covered by this test cast `'{"method": "PUT", "path": "users-service/users/<payload>", "bod": "<payload>"},` since `payload` is a dictionary and it is set in the middle part in the string `"users-service/users/<payload>"`

[the remark this answers is not in the document]
Comment (maxhoheiser) [reply localstack-core/localstack/services/events/target.py:96]: This will interestingly fail for AWS since it will not be correctly transformed since it would result in a string that is not a json string but the system expects a json string

I added a test for this failure also

File: tests/aws/services/events/test_events_inputs.py [hunk 3]
```diff
@@ -462,3 +462,55 @@ def test_input_transformer_predefined_variables(
...
+    @pytest.mark.skipif(
+        is_old_provider(),
+        reason="V1 provider does not support this feature",
+    )
+    @pytest.mark.parametrize(
+        "input_template",
+        [
+            '{"method": "PUT", "path": "users-service/users/<userId>", "bod": <payload>}',
+            '"Payload of <payload> with path users-service/users/<userId>"',
+        ],
```
[the remark this answers is not in the document]
Comment (maxhoheiser) [reply tests/aws/services/events/test_events_inputs.py]: added something similar to as a test case also with nested lists and target replacement values

File: localstack-core/localstack/services/events/utils.py (hunk 1 above)
Comment (maxhoheiser) [reply localstack-core/localstack/services/events/utils.py:278]: I added a test case for nested dicts in string transformer `'{"method": "PUT", "nested": {"level1": {"level2": {"level3": "users-service/users/<userId>"} } }, "bod": "<userId>"}',`

File: localstack-core/localstack/services/events/utils.py [hunk 4]
```diff
@@ -248,3 +248,32 @@ def get_trace_header_encoded_region_account(
             return json.dumps({"original_id": original_id, "original_account": source_account_id})
         else:
             return json.dumps({"original_account": source_account_id})
+
+
+def is_nested_in_string(template, match) -> bool:
```
[the remark this answers is not in the document]
Comment (maxhoheiser) [reply localstack-core/localstack/services/events/utils.py]: added unit test for the function `test_is_nested_in_string`

[the remark this answers is not in the document]
Comment (bentsku) [reply localstack-core/localstack/services/events/utils.py]: Awesome, thanks! I'll wait for the pipeline to be green and review the changes. Thank you!

File: localstack-core/localstack/services/events/target.py [hunk 5]
```diff
@@ -75,23 +76,50 @@ def get_template_replacements(
...
+            return loaded_json_template
+        except json.JSONDecodeError:
+            LOG.info(
+                json.dumps(
+                    {
+                        "InfoCode": "InternalInfoEvents at transform_event",
+                        "InfoMessage": f"Replaced template is not valid json: {formatted_template}",
+                    }
+                )
+            )
```
Comment (bentsku) [inline localstack-core/localstack/services/events/target.py:120]: nit: I'm not sure I see the value of JSON dumping a log at the `INFO` level here. Right now, our logs are used by humans, so reading JSON might not be the simplest? Should it be a simpler sentence?

**Review by bentsku — APPROVED** [review]
Wow, quite a lot of coverage now! Thank you for addressing the comments, I know it was a lot 😅 I think this is really great, thanks for fixing this issue! This was quite the rabbit hole... 
I only have a small nit regarding the log statement dumping JSON. 

I've done a run for -ext, and it was all green, so we should be on the safe side! LGTM! And congrats for this quite big addition! 😄

File: localstack-core/localstack/services/events/target.py (hunk 5 above)
Comment (maxhoheiser) [reply localstack-core/localstack/services/events/target.py:120]: we need this info for event studio, I know its not the most elegant but this is what we have right now to capture specific internal conditions and display them to users in a structured way in event studio
